datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,138 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command and function calls, allowing for dry runs and output handling
+
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any
+
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class ReaderThread(threading.Thread):
+    def __init__(self, file, q):
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = q
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        os.read() or queue.put(). We are closing the file that read
+        is reading from here, but the queue has to be emptied in
+        another thread in order to ensure thread-exiting.
+        """
+        self.quit = True
+        self.file.close()
+
+    def run(self):
+        logger.debug(f"ReaderThread({self.file}, {self.queue}) started")
+        while not self.quit:
+            try:
+                data = os.read(self.file.fileno(), 1024)
+            except BrokenPipeError as exc:
+                logger.debug(f"{self} exiting (broken pipe)")
```

Comment:
although just a minor opt, in general - avoid unconditional formatting here and elsewhere of log messages -- delay until/if line is actually logged:

```suggestion
                logger.debug("%s exiting (broken pipe)", self)
```